File: mamba2-burn/examples/debug_mixer.py
#!/usr/bin/env python3
"""Debug Mamba2Mixer operations in detail."""

import torch
import numpy as np
import json
from transformers import AutoTokenizer, AutoModelForCausalLM

def capture_tensor_info(name, tensor):
    """Capture detailed tensor information."""
    if isinstance(tensor, torch.Tensor):
        data = tensor.detach().cpu().float().numpy()
        
        return {
            "name": name,
            "shape": list(data.shape),
            "mean": float(np.mean(data)),
            "std": float(np.std(data)),
            "min": float(np.min(data)),
            "max": float(np.max(data)),
            "first_10": data.flat[:10].tolist(),
            "abs_mean": float(np.mean(np.abs(data))),
        }
    return None

def hook_mixer_internals(mixer, layer_idx, outputs):
    """Add hooks to capture mixer internal operations."""
    captured = outputs[f"layer_{layer_idx}_mixer"] = {}
    
    def capture_hook(name):
        def hook(module, input, output):
            if isinstance(input, tuple):
                input = input[0]
            if isinstance(output, tuple):
                output = output[0]
            captured[f"{name}_input"] = capture_tensor_info(f"{name}_input", input)
            captured[f"{name}_output"] = capture_tensor_info(f"{name}_output", output)
        return hook
    
    # Hook input projection
    if hasattr(mixer, 'in_proj'):
        mixer.in_proj.register_forward_hook(capture_hook("in_proj"))
    
    # Hook convolution
    if hasattr(mixer, 'conv1d'):
        mixer.conv1d.register_forward_hook(capture_hook("conv1d"))
    
    # Hook activation
    if hasattr(mixer, 'act'):
        mixer.act.register_forward_hook(capture_hook("act"))
    
    # Hook norm
    if hasattr(mixer, 'norm'):
        mixer.norm.register_forward_hook(capture_hook("norm"))
    
    # Hook output projection
    if hasattr(mixer, 'out_proj'):
        mixer.out_proj.register_forward_hook(capture_hook("out_proj"))

def main():
    outputs = {}
    
    # Setup
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    print(f"Using device: {device}")
    
    # Load model
    model_name = "AntonV/mamba2-130m-hf"
    tokenizer = AutoTokenizer.from_pretrained(model_name)
    model = AutoModelForCausalLM.from_pretrained(
        model_name,
        torch_dtype=torch.float32
    ).to(device)
    model.eval()
    
    # Prepare input
    prompt = "Hey how are you doing?"
    inputs = tokenizer(prompt, return_tensors="pt").to(device)
    input_ids = inputs["input_ids"]
    
    # Get embeddings
    with torch.no_grad():
        embeddings = model.backbone.embeddings(input_ids)
        outputs["embeddings"] = capture_tensor_info("embeddings", embeddings)
        
        # Process first layer only
        layer = model.backbone.layers[0]
        
        # Hook mixer internals
        hook_mixer_internals(layer.mixer, 0, outputs)
        
        # Forward through layer
        hidden_states = embeddings
        outputs["layer_0_input"] = capture_tensor_info("layer_0_input", hidden_states)
        
        # Norm
        normed = layer.norm(hidden_states)
        outputs["layer_0_norm"] = capture_tensor_info("layer_0_norm", normed)
        
        # Mixer (will trigger hooks)
        mixer_output = layer.mixer(normed)[0]
        outputs["layer_0_mixer_output"] = capture_tensor_info("layer_0_mixer_output", mixer_output)
        
        # Final output
        output = mixer_output + hidden_states
        outputs["layer_0_output"] = capture_tensor_info("layer_0_output", output)
    
    # Save outputs
    with open("python_mixer_debug.json", 'w') as f:
        json.dump(outputs, f, indent=2)
    
    print("\nCaptured tensors:")
    for key in sorted(outputs.keys()):
        if isinstance(outputs[key], dict) and 'mean' in outputs[key]:
            info = outputs[key]
            print(f"{key}: mean={info['mean']:.6f}, std={info['std']:.6f}")
    
    # Also print mixer internals
    if "layer_0_mixer" in outputs:
        print("\nMixer internals:")
        for key in sorted(outputs["layer_0_mixer"].keys()):
            info = outputs["layer_0_mixer"][key]
            if info:
                print(f"  {info['name']}: mean={info['mean']:.6f}, std={info['std']:.6f}")

if __name__ == "__main__":
    main()

File: mamba2-burn/examples/debug_mixer.rs
use anyhow::Result;
use burn::tensor::{Tensor, Int};
use mamba2_burn::prelude::*;
use serde_json::{json, Value};
use std::collections::HashMap;
use std::fs::File;
use tokenizers::Tokenizer;

// Type alias for backend
#[cfg(feature = "tch-gpu")]
type Backend = burn::backend::LibTorch;
#[cfg(not(feature = "tch-gpu"))]
type Backend = burn::backend::NdArray;

fn capture_tensor_info<B: burn::prelude::Backend, const D: usize>(
    name: &str,
    tensor: &Tensor<B, D>
) -> Value {
    let data = tensor.clone().to_data();
    let shape: Vec<usize> = data.shape.clone();
    let values: Vec<f32> = data.convert::<f32>().as_slice::<f32>().unwrap().to_vec();
    
    let mean = values.iter().sum::<f32>() / values.len() as f32;
    let variance = values.iter().map(|&x| (x - mean).powi(2)).sum::<f32>() / values.len() as f32;
    let std = variance.sqrt();
    let min = values.iter().fold(f32::INFINITY, |a, &b| a.min(b));
    let max = values.iter().fold(f32::NEG_INFINITY, |a, &b| a.max(b));
    let abs_mean = values.iter().map(|x| x.abs()).sum::<f32>() / values.len() as f32;
    
    json!({
        "name": name,
        "shape": shape,
        "mean": mean,
        "std": std,
        "min": min,
        "max": max,
        "first_10": &values[..10.min(values.len())],
        "abs_mean": abs_mean,
    })
}

fn main() -> Result<()> {
    let mut outputs = HashMap::new();
    let device = auto_device();
    println!("Using device: {:?}", device);
    
    // Load model
    let model_name = "AntonV/mamba2-130m-hf";
    let cache_dir = std::env::var("HF_HOME").unwrap_or_else(|_| {
        format!("{}/.cache/huggingface", std::env::var("HOME").unwrap())
    });
    let model_path = format!("{}/hub/models--{}", cache_dir, model_name.replace("/", "--"));
    let snapshots_dir = format!("{}/snapshots", model_path);
    let entries: Vec<_> = std::fs::read_dir(&snapshots_dir)?
        .filter_map(|e| e.ok())
        .filter(|e| e.file_type().ok().map(|ft| ft.is_dir()).unwrap_or(false))
        .collect();
    let snapshot_path = entries[0].path();
    
    // Load tokenizer and model
    let tokenizer_path = snapshot_path.join("tokenizer.json");
    let tokenizer = Tokenizer::from_file(&tokenizer_path)
        .map_err(|e| anyhow::anyhow!("Failed to load tokenizer: {:?}", e))?;
    
    let (_config, model) = load_mamba2_weights::<Backend>(&snapshot_path, &device)?;
    
    // Prepare input
    let prompt = "Hey how are you doing?";
    let encoding = tokenizer.encode(prompt, false)
        .map_err(|e| anyhow::anyhow!("Failed to encode: {:?}", e))?;
    let input_ids: Vec<i64> = encoding.get_ids().iter().map(|&id| id as i64).collect();
    let input_tensor = Tensor::<Backend, 1, Int>::from_data(
        input_ids.as_slice(),
        &device
    ).reshape([1, input_ids.len()]);
    
    // Get embeddings
    let embeddings = model.model.embeddings.forward(input_tensor);
    outputs.insert("embeddings".to_string(), capture_tensor_info("embeddings", &embeddings));
    
    // Process first layer
    let layer = &model.model.layers[0];
    let hidden_states = embeddings;
    outputs.insert("layer_0_input".to_string(), capture_tensor_info("layer_0_input", &hidden_states));
    
    // Norm
    let normed = layer.norm.forward(hidden_states.clone());
    outputs.insert("layer_0_norm".to_string(), capture_tensor_info("layer_0_norm", &normed));
    
    // Mixer - we need to capture internal operations
    // For now, just capture input and output
    let mixer_output = layer.mixer.forward(normed.clone(), None, 0);
    outputs.insert("layer_0_mixer_output".to_string(), capture_tensor_info("layer_0_mixer_output", &mixer_output));
    
    // To debug mixer internals, we need to manually reproduce the mixer forward pass
    let mixer = &layer.mixer;
    let mut mixer_internals = HashMap::new();
    
    // Input projection
    let in_proj_output = mixer.in_proj.forward(normed.clone());
    mixer_internals.insert("in_proj_output".to_string(), 
        capture_tensor_info("in_proj_output", &in_proj_output));
    
    // Get dimensions for splitting
    let [batch, seq_len, _] = normed.dims();
    let d_inner = mixer.d_inner;
    let _chunks = mixer.chunk_size;
    
    // Split projections
    let proj_size = in_proj_output.dims()[2];
    let dt_start = d_inner * 2 + 2 * mixer.n_groups * mixer.d_state;
    
    let xbc = in_proj_output.clone().slice([0..batch, 0..seq_len, 0..(d_inner * 2 + 2 * mixer.n_groups * mixer.d_state)]);
    let dt = in_proj_output.slice([0..batch, 0..seq_len, dt_start..proj_size]);
    
    mixer_internals.insert("xbc_slice".to_string(), 
        capture_tensor_info("xbc_slice", &xbc));
    mixer_internals.insert("dt_slice".to_string(), 
        capture_tensor_info("dt_slice", &dt));
    
    // Final output
    let output = mixer_output + hidden_states;
    outputs.insert("layer_0_output".to_string(), capture_tensor_info("layer_0_output", &output));
    
    // Add mixer internals to outputs
    outputs.insert("layer_0_mixer".to_string(), json!(mixer_internals));
    
    // Save outputs
    let file = File::create("rust_mixer_debug.json")?;
    serde_json::to_writer_pretty(file, &outputs)?;
    
    println!("\nCaptured tensors:");
    let mut keys: Vec<_> = outputs.keys().cloned().collect();
    keys.sort();
    for key in keys {
        if let Some(info) = outputs.get(&key) {
            if let Some(mean) = info.get("mean") {
                if let Some(std) = info.get("std") {
                    println!("{}: mean={:.6}, std={:.6}", 
                        key,
                        mean.as_f64().unwrap_or(0.0),
                        std.as_f64().unwrap_or(0.0)
                    );
                }
            }
        }
    }
    
    Ok(())
}